src/clib/pio_spmd.c

```diff
@@ -328,134 +370,162 @@ int pio_swapm(void *sndbuf,   int sndlths[], int sdispls[],  MPI_Datatype stypes
-        }else{
+        }
+	else
+	{
             maxreq = 2;
             maxreqh = 1;
         }
     }
-    if(handshake){
+
+    /* If handshaking is in use, do a nonblocking recieve to listen
+     * for it. */
+    if (handshake)
+    {
         hs = 1;
-        for(istep=0; istep<maxreq; istep++){
+        for (istep = 0; istep < maxreq; istep++)
+	{
             p = swapids[istep];
-            if( sndlths[p] > 0){
-                tag = mytask+offset_t;
-                CheckMPIReturn(MPI_Irecv( &hs, 1, MPI_INT, p, tag, comm, hs_rcvids+istep), __FILE__,__LINE__);
+            if (sndlths[p] > 0)
+	    {
+                tag = mytask + offset_t;
+                CheckMPIReturn(MPI_Irecv(&hs, 1, MPI_INT, p, tag, comm, hs_rcvids + istep),
+			       __FILE__,__LINE__);
             }
         }
     }
-    for(istep=0;istep < maxreq; istep++){
+    
+    for (istep = 0; istep < maxreq; istep++)
+    {
         p = swapids[istep];
-        if(rcvlths[p] > 0){
+        if (rcvlths[p] > 0)
+	{
             tag = p + offset_t;
-            ptr = (void *)((char *) rcvbuf + rdispls[p]);
+            ptr = (void *)((char *)rcvbuf + rdispls[p]);
 
-            //    printf("%s %d %d %d\n",__FILE__,__LINE__,p,(int) rtypes[p]);
-            CheckMPIReturn(MPI_Irecv( ptr, rcvlths[p], rtypes[p], p, tag, comm, rcvids+istep), __FILE__,__LINE__);
+            CheckMPIReturn(MPI_Irecv(ptr, rcvlths[p], rtypes[p], p, tag, comm, rcvids+istep),
+			   __FILE__,__LINE__);
 
-            if(handshake)
-                CheckMPIReturn(MPI_Send( &hs, 1, MPI_INT, p, tag, comm), __FILE__,__LINE__);
+            if (handshake)
+                CheckMPIReturn(MPI_Send(&hs, 1, MPI_INT, p, tag, comm), __FILE__,__LINE__);
         }
     }
 
+    /* ??? */
     rstep = maxreq;
-    for(istep = 0; istep < steps; istep++){
+    for (istep = 0; istep < steps; istep++)
+    {
         p = swapids[istep];
-        if(sndlths[p] > 0){
+        if (sndlths[p] > 0)
+	{
             tag = mytask + offset_t;
-            if(handshake){
-                CheckMPIReturn(MPI_Wait ( hs_rcvids+istep, &status), __FILE__,__LINE__);
+            if (handshake)
+	    {
+                CheckMPIReturn(MPI_Wait(hs_rcvids+istep, &status), __FILE__,__LINE__);
                 hs_rcvids[istep] = MPI_REQUEST_NULL;
             }
-            ptr = (void *)((char *) sndbuf + sdispls[p]);
+            ptr = (void *)((char *)sndbuf + sdispls[p]);
 
-            if(isend){
-                CheckMPIReturn(MPI_Irsend(ptr, sndlths[p], stypes[p], p, tag, comm,sndids+istep), __FILE__,__LINE__);
-            }else{
+            if (isend)
+                CheckMPIReturn(MPI_Irsend(ptr, sndlths[p], stypes[p], p, tag, comm, sndids + istep),
+			       __FILE__,__LINE__);
+	    else
                 CheckMPIReturn(MPI_Send(ptr, sndlths[p], stypes[p], p, tag, comm), __FILE__,__LINE__);
-            }
-
         }
-        if(istep > maxreqh){
+
+	/* ??? */
+        if (istep > maxreqh)
```

Comment:
we did comms in sets of size max_reqs, if istep > maxreqh then there is a remainder that must be handled.